Add User-Adjustable Object Tracing Parameters (plumbed to existing arguments in the code)

tbowers7 opened the pull request:
### Overview

No new base functionality is being added in the PR, but two object tracing function arguments have been plumbed upward to user-settable parameters (called `trace_maxshift` and `trace_min_max`), and two other `FindObj` parameters have been renamed to indicate their function in object _tracing_ rather than _finding_.

**NOTE:** The `FindObj` parameter name change is a breaking change, but involves two somewhat obscure parameters (one of which is not actually consumed by the code at the moment) and no changes were required for DevSuite canned `.pypeit` files.  So, maybe only a "bending" change?

### Background

This set of changes was triggered by trying to make PypeIt follow the trace of low-dispersion spectra taken at high airmass.  The consequence of this observational choice is that the object trace is very noticeably curved on the detector due to atmospheric dispersion.

For example, shown below are the `spec2d` file and object tracing QA plot for an object observed with LDT/DeVeny using the DV1 (150 g/mm) grating at an elevation of 11º above the horizon (airmass 5).

<img width="530" height="600" alt="ginga pypeit_show_2dspec" src="https://github.com/user-attachments/assets/0d289d24-8d2e-4173-ba9d-0c1cce5add3c" />
<img width="435" height="600" alt="objtraceQA" src="https://github.com/user-attachments/assets/c5248ab0-76e3-457b-a0c3-bb754a82aba2" />


At blue wavelengths, the trace curves faster than the algorithm can keep up, and at red wavelengths the trace starts to follow 2nd order light at a location where 1st order light is too red for the silicon detector.  This reduction made use of `FindObj` parameter `find_min_max = 900,1700` to identify the star and `find_numiterfit = 100` to iterate the number of attempts at tracing the star.

### Solution

The solution was to modify the `maxshift` argument to the function `pypeit.core.trace.fit_trace()` to be larger than 1 pixel.  To do this required plumbing the argument chain back high enough to reach a function that knew about user parameters and defining a new `FindObj` parameter called `trace_maxshift`.

Additionally, I found it useful to plumb upward the argument `spec_min_max` from `pypeit.core.findobj_skymask.objs_in_slit()` to a new parameter `trace_min_max`, which specifies the spectral range *in pixels* over which to attempt to trace the object.  This is useful for instances where the selected slit/order would include regions without expected signal (_e.g._ bluer than the atmospheric cutoff or redder than the silicon cutoff).  **NOTE:** This is distinct from the parameter `find_min_max` that limits the range over which the 2D spectrum is flattened for object finding.

In the images below, the following `FindObj` parameters were used: `find_min_max = 900,1700`, `find_numiterfit = 100`, `trace_maxshift = 2.0`, and `trace_min_max = 100,1700`, where the final parameter corresponds to a wavelength range from about 3300Å to 10,200Å.

<img width="530" height="600" alt="Screenshot 2025-10-24 at 1 46 07 PM" src="https://github.com/user-attachments/assets/dfa03aab-58d1-491e-bc7f-9ee5a80fc9f0" />
<img width="435" height="600" alt="objtraceQA" src="https://github.com/user-attachments/assets/29e253f4-e0bd-43b7-9ae8-22681876ad23" />

The trace now easily follows the star and ignores the 2nd order light at the far red end of the detector.

The updated base parameters are included in the LDT/DeVeny spectrograph class.

### Additional Changes

#### Object Tracing QA Plot

To keep an eye on the object trace as a function of spectral position, I found it convenient to create a `objtrace_QA()` function that used debugging code from the aforementioned `fit_trace()` function.  The (re-)introduction of an object trace QA then inadvertently reactivated some HTML-generation code in `pypeit.core.qa`, so naturally I worked to get all of that working again nicely.  The additional QA plot, adjustments to the QA filename generator & HTML pages, and conformal changes (primarily in the spectral flexure code) make up the bulk of this PR.

#### `FindObj` Parameter Changes

Along the lines of separating object _finding_ from object _tracing_, two parameters were renamed to indicate their use in object tracing: `find_maxdev` -> `trace_maxdev` and `find_extrap_npoly` -> `trace_extrap_npoly`, the latter of which is not actually consumed by the code at this time.

### DevSuite Runs

A run of the DevSuite started on 21 October ran successfully (`tbowers-objfind-21ct.report`):
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  275 passed, 465 warnings in 421.41s (0:07:01) ---
--- PYTEST UNIT TESTS PASSED  157 passed, 93617 warnings in 929.09s (0:15:29) ---
--- PYTEST VET TESTS PASSED  69 passed, 139915 warnings in 5494.65s (1:31:34) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 281/281 TESTS  ---
Testing Started at 2025-10-22T01:51:51.050956
Testing Completed at 2025-10-22T17:33:14.742730
Total Time: 15:41:23.691774
```

~~Another run is in progress that includes much of the HTML-related code work done since the initial run.~~
Complete run passed 10/28/25 -- see comment below.

tbowers7 committed:
I think I was doing something here...

But I can't remember anymore exactly what it was...

This branched off 1.15.0

	modified:   pypeit/coadd2d.py
	modified:   pypeit/core/findobj_skymask.py
	modified:   pypeit/core/qa.py
	modified:   pypeit/core/trace.py
	modified:   pypeit/find_objects.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Add slitless_pixflat to excluded frametypes

tbowers7 committed:
Limit wavelength solution polyorder for DV6+

A 4th order polynomial does no better than a 3rd order for the 600 g/mm+
gratings.

	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Fix coadd2d output filenames to remove spaces

This is standard for regular output files; make it the same for
coadd2d.

	modified:   pypeit/coadd2d.py

tbowers7 committed:
Debugging...

	modified:   pypeit/core/findobj_skymask.py
	modified:   pypeit/core/qa.py
	modified:   pypeit/par/pypeitpar.py

tbowers7 committed:
Parameter renaming to indicate usage

``find_<>`` renamed to ``trace_<>`` where appropriate.

	modified:   pypeit/coadd2d.py
	modified:   pypeit/core/findobj_skymask.py
	modified:   pypeit/core/qa.py
	modified:   pypeit/find_objects.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Add new FindObj par ``trace_min_max``

This parameter feeds the extant argument in ``findobj_skymask.objs_in_slit()``
called ``spec_min_max`` that limits the spectral range (measured in pixels)
over which the trace is fix.  This is useful for situations where the spectral
range on the detector exceeds the region where sky photons can be detected
(either atmospheric or silicon cutoffs or both).  The default value is ``None``
which corresponds to previous behavior of using the entire chip.

Conformal changes.

	modified:   pypeit/coadd2d.py
	modified:   pypeit/core/findobj_skymask.py
	modified:   pypeit/find_objects.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Docstrings & QA HTML pages

Oh my!

	modified:   pypeit/core/findobj_skymask.py
	modified:   pypeit/core/qa.py

tbowers7 committed:
Squeaky Clean

Undo superfluous edits.

	modified:   pypeit/coadd2d.py
	modified:   pypeit/core/findobj_skymask.py
	modified:   pypeit/core/flexure.py
	modified:   pypeit/core/qa.py
	modified:   pypeit/core/trace.py
	modified:   pypeit/par/pypeitpar.py

tbowers7 committed:
Make HTML pages work for spec flex QA

Need to pass around bits of the filename structure.

	modified:   pypeit/core/flexure.py
	modified:   pypeit/par/pypeitpar.py

tbowers7 committed:
Merge branch 'develop' into deveny_objtrace

tbowers7 committed:
Exercise in typing

	modified:   pypeit/core/findobj_skymask.py
	modified:   pypeit/core/flexure.py
	modified:   pypeit/core/qa.py

tbowers7 committed:
Fix QA paths

	modified:   pypeit/core/qa.py

tbowers7 committed:
Remove type hints for `specobjs` in core

This is indicative of a larger issue we'll need to address in PypeIt:
ensuring the `core` routines do not depend on PypeIt-specific classes,
but rather only base Python / NumPy / AstroPy objects.

	modified:   pypeit/core/flexure.py

tbowers7 committed:
Merge branch 'develop' into deveny_objtrace

tbowers7 committed:
Merge branch 'develop' into deveny_objtrace

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -1417,6 +1417,152 @@ def objfind_QA(spat_peaks, snr_peaks, spat_vector, snr_vector, snr_thresh, qa_ti
     plt.close('all')
 
+
+def objtrace_QA(
```
Nice!  Thanks for isolating this.  Can we edit `fit_trace` in `trace.py` so that it uses this function?

tbowers7 replied:
Not unless we move `objtrace_QA()` somewhere else because of circular import errors `pypeit.core.trace` <--> `pypeit.core.findobj_skymask`

kbwestfall replied:
Okay, makes sense.  Happy to leave it as is.

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -1417,6 +1417,152 @@ def objfind_QA(spat_peaks, snr_peaks, spat_vector, snr_vector, snr_thresh, qa_ti
+
+    # Construct boolean flags
+    inpgpm = np.invert(_trace_bpm)
+    cengpm = np.invert(msk.astype(bool))
```
- We tend to prefer `logical_not` over `invert` (although we've never done a full find/replace).
- Given that `msk` is only used to construct `cengpm`, we might consider making `cengpm` the argument, instead of `msk`.  I.e., force the conversion from `msk` to `cengpm` to be done by the calling function.

tbowers7 replied:
This function has been updated for ``np.invert`` --> ``np.logical_not``.  I will submit a separate PR that does the full-repo find/replace for both that and ``np.in1d`` --> ``np.isin``, because this latter change will be required upon the release of NumPy 2.4 in the coming months.

tbowers7 replied:
Changed `msk` -> `cen_gpm` for the input.

kbwestfall replied:
On converting `np.in1d` -> `np.isin`, note that this is already done in #2007 .

tbowers7 replied:
Just did the ``np.invert`` --> ``np.logical_not`` as a PR **_into_** #2007.

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -1417,6 +1417,152 @@ def objfind_QA(spat_peaks, snr_peaks, spat_vector, snr_vector, snr_thresh, qa_ti
+    msk
+        Measurement flags with a data type depending on ``bitmask``
+    trace_cen
+        Input trace spatial pixel centroids as a function of spectral pixel.
```
Some of the differences between the arguments are hard to understand.  I.e., what's the difference between `cen` and `trace_cen`?  Can you make this more clear?  And maybe it would be helpful to comment on the expected shapes of the provided arrays, particularly which ones are expected to match.

tbowers7 replied:
Parameter names and docstring descriptions updated.

tbowers7 committed:
Merge branch 'develop' into deveny_objtrace

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -1417,6 +1417,152 @@ def objfind_QA(spat_peaks, snr_peaks, spat_vector, snr_vector, snr_thresh, qa_ti
+    if show:
+        plt.show()
+    if objtraceQA_filename is not None:
+        fig.savefig(objtraceQA_filename, dpi=400)
```
Does this work if `show` is True and the file name is not None?  FWIW, I tend to treat these as mutually exclusive by only having one parameter:
```
if objtraceQA_filename is None:
    plt.show()
else:
    ...
fig.clear()
plt.close(fig)
```

tbowers7 replied:
The way this is currently written, there are four possible outcomes (which I think is correct):
1. `show = False`, `objtraceQA_filename is None`: No plot is shown to screen or written to disk.  (likely not to happen)
2. `show = True`, `objtraceQA_filename is None`: Plot is shown on screen, no plot is written to disk.  (likely not to happen)
3. `show = False`, `objtraceQA_filename is not None`: No plot is shown on screen, QA written to disk.  (likely standard use case for most `run_pypeit` reductions)
4. `show = True`, `objtraceQA_filename is not None`: Plot is shown on screen, QA written to disk.  (likely to be the case when `-s` is passed to `run_pypeit`)

I'll defer to @kbwestfall for guidance on how this fits into the larger QA ecosystem for PypeIt.

debora-pe replied:
I think this looks good, and it's the same as the `objfind_QA()`

kbwestfall replied:
Sounds good.  I'm deferring to @debora-pe on this!

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -1924,13 +2073,14 @@ def objs_in_slit(image, ivar, thismask, slit_left, slit_righ,
         spec_mask = (spec_vec >= spec_min_max_out[0]) & (spec_vec <= spec_min_max_out[1])
         trc_inmask = np.outer(spec_mask, np.ones(len(sobjs), dtype=bool))
-        xfit_fweight = fit_trace(image, xinit_fweight, ncoeff, bpm=np.invert(inmask), maxshift=1., niter=numiterfit,
+        # trace_fit, cen, err, msk, traceset = thing1
```
What does this comment mean?

tbowers7 replied:
Absolutely nothing.  A commented line when I was trying to figure out what the outputs of `fit_trace()` are.  Gone.

kbwestfall commented on pypeit/core/flexure.py:
```diff
@@ -1158,7 +1156,24 @@ def get_sky_spectrum(sciimg, ivar, waveimg, thismask, global_sky, box_radius, sl
+        particular object, trace, or location of interest.
+    name
+        Object, trace, or location name to be printed in the plot
+    """
```
Thanks for the added docs!

kbwestfall commented on pypeit/core/flexure.py:
```diff
@@ -1209,12 +1225,14 @@ def spec_flexure_qa(slitords, bpm, basename, flex_list,
             is used.
     """
+    # Extract the mode and detector from the ``basename``
+    *_, mode, det = basename.split("_")
```
Nice!  I didn't know about the use of `*_`.

tbowers7 replied:
Yes, it's very nice-uh.

kbwestfall commented on pypeit/core/flexure.py:
```diff
@@ -1209,12 +1225,14 @@ def spec_flexure_qa(slitords, bpm, basename, flex_list,
     # What type of QA are we doing
-    slit_cen = False
-    if specobjs is None: slit_cen = True
+    slit_cen = True if specobjs is None else False
```
`slit_cen = specobjs is None`

tbowers7 replied:
🤦‍♂️

kbwestfall commented on pypeit/par/pypeitpar.py:
```diff
@@ -4351,13 +4351,27 @@ def __init__(self, trace_npoly=None, snr_thresh=None, find_trim_edge=None,
+        defaults['trace_maxdev'] = 2.0
+        dtypes['trace_maxdev'] = [int, float]
+        descr['trace_maxdev'] = 'Maximum deviation of pixels from polynomial fit to trace used to reject bad pixels in trace fitting.  (For ``pypeit<=1.18.x``, this ' \
+                                      'parameter was called ``find_maxdev``.)'
```
I agree that these parameter name changes make sense, but it's worth noting that they are breaking changes.

tbowers7 replied:
Noted in the preamble to this PR and added label.  And noted in the Release Notes.

kbwestfall commented on pypeit/find_objects.py:
```diff
@@ -690,6 +690,19 @@ def __repr__(self):
         txt += '>'
         return txt
 
+    def get_findobj_qa_filename(self, slit:int, neg:bool, save_objfindQA:bool) -> str:
```
Nice!

kbwestfall reviewed (changes requested):
These look good to me!  I'm "requesting changes," but they're all pretty minor.  Are there any dev-suite updates necessary given the changes to parameter names?

tbowers7 committed:
Add TODO note about `core` functions

Can probably undo circular import errors like annotating
`specobjs` in `core` functions with Python 3.14, but really we
shouldn't be using PypeIt-specific objects in `core` to begin
with.

	modified:   pypeit/core/flexure.py

tbowers7 commented:
> These look good to me! I'm "requesting changes," but they're all pretty minor. Are there any dev-suite updates necessary given the changes to parameter names?

No dev-suite updates necessary... no one is using `find_maxdev` (or the currently useless `find_extrap_npoly`, for that matter) in any of the canned PypeIt files.

tbowers7 commented:
DevSuite tests pass (`tbowers-objfind-27oct.report`):
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  275 passed, 471 warnings in 394.44s (0:06:34) ---
--- PYTEST UNIT TESTS PASSED  157 passed, 93527 warnings in 887.98s (0:14:47) ---
--- PYTEST VET TESTS PASSED  69 passed, 141288 warnings in 4968.31s (1:22:48) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 281/281 TESTS  ---
Testing Started at 2025-10-27T14:59:02.253858
Testing Completed at 2025-10-28T04:50:29.522674
Total Time: 13:51:27.268816
```

tbowers7 committed:
Low-hanging PR fruit

	modified:   pypeit/core/findobj_skymask.py
	modified:   pypeit/core/flexure.py

tbowers7 committed:
Remaining PR comments from KBW

	modified:   pypeit/core/findobj_skymask.py

tbowers7 committed:
Merge branch 'develop' into deveny_objtrace

tbowers7 committed:
PR Release Notes

	modified:   doc/releases/1.18.2dev.rst

tbowers7 committed:
pedantic

adjective

of or like a pedant.

pedant

noun

a person who is excessively concerned with minor details and rules or with
displaying academic learning.

	modified:   pypeit/core/findobj_skymask.py

tbowers7 committed:
Merge branch 'develop' into deveny_objtrace

tbowers7 committed:
Merge fixes

	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Merge branch 'develop' into deveny_objtrace

debora-pe reviewed (approved):
@tbowers7 sorry for taking so long to review this! 
It looks good to me. I would ask 2 minor things if you are willing to spend a little more time. Can you add a small description of the new plot in the docs (like [here](https://pypeit.readthedocs.io/en/release/qa.html#exposure-qa)). I feel like it can be hard for a common user to understand it.
Also, I have tested on a dataset that has 5 objects detected and the plot looked very crowded. Do you mind taking a look at this aspect and find a way to make it mostly legible in various cases. If it is helpful, the dataset I tested it on is `keck_lris_red_mark4/long_400_8500_d560`

I'm approving anyway now, but I hope you will consider these small comments. Thanks!

tbowers7 commented:
>  I would ask 2 minor things if you are willing to spend a little more time. Can you add a small description of the new plot in the docs (like [here](https://pypeit.readthedocs.io/en/release/qa.html#exposure-qa)). I feel like it can be hard for a common user to understand it. Also, I have tested on a dataset that has 5 objects detected and the plot looked very crowded. Do you mind taking a look at this aspect and find a way to make it mostly legible in various cases. If it is helpful, the dataset I tested it on is `keck_lris_red_mark4/long_400_8500_d560`

Yes, I will make these changes.  Thanks for your review!

kbwestfall reviewed (approved):
Thanks, @tbowers7 !  Clearing this to be merged, but we should wait to do that until after you've addressed @debora-pe requests.

tbowers7 commented:
> Thanks, @tbowers7 ! Clearing this to be merged, but we should wait to do that until after you've addressed @debora-pe requests.

Hi @kbwestfall .  Thanks for the clear.  Yes, I'd like to address @debora-pe 's requests -- will do in the next couple days.

tbowers7 committed:
Adding documentation for the ObjTrace QA

	new file:   doc/figures/objtrace_high_airmass_bad.png
	new file:   doc/figures/objtrace_high_airmass_good.png
	new file:   doc/figures/objtrace_qa_highX_bad.png
	new file:   doc/figures/objtrace_qa_highX_good.png
	modified:   doc/object_finding.rst
	modified:   doc/qa.rst

tbowers7 committed:
Update to docs surrounding object tracing

Also includes fixes to the object finding and tracing QA plots
following the request of @debora-pe .

	modified:   doc/figures/deveny_findobj_qa.png
	new file:   doc/figures/deveny_traceobj_qa.png
	new file:   doc/figures/gmos_objfind_qa_example1.png
	new file:   doc/figures/gmos_objfind_qa_example2.png
	new file:   doc/figures/gmos_objfind_qa_example3.png
	new file:   doc/figures/lris_objfind_qa_example.png
	new file:   doc/figures/lris_objtrace_qa_example.png
	modified:   doc/object_finding.rst
	modified:   doc/qa.rst
	modified:   doc/spectrographs/ldt_deveny.rst
	modified:   pypeit/core/findobj_skymask.py

tbowers7 commented:
Hi @debora-pe .  I've added the requested description to the documentation.  How does the revamped QA plot look for your case of ``keck_lris_red_mark4/long_400_8500_d560``?

<img width="800" height="600" alt="lris_objtrace_qa_example" src="https://github.com/user-attachments/assets/07bfc16e-f765-4385-9982-9a70d17ce3bd" />

If all looks good, this PR is ready for merge @kbwestfall .

debora-pe commented:
> Hi @debora-pe . I've added the requested description to the documentation. How does the revamped QA plot look for your case of `keck_lris_red_mark4/long_400_8500_d560`?

@tbowers7 This is all great! Thank you so much for doing it. I just notice that file "gmos_objtrace_qa_example.png" is missing. Besides that, the PR can be merged.

tbowers7 committed:
Missing QA file for the docs

	new file:   doc/figures/gmos_objtrace_qa_example.png
